src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs

```diff
@@ -278,43 +318,101 @@ private void ProcessConnectionByTCPPort(string targetNameOrAddress)
 
             WriteObject(false);
         }
+
         #endregion ConnectionTest
 
         #region TracerouteTest
+
         private void ProcessTraceroute(string targetNameOrAddress)
         {
             byte[] buffer = GetSendBuffer(BufferSize);
 
-            string resolvedTargetName;
-            IPAddress targetAddress;
-            if (!InitProcessPing(targetNameOrAddress, out resolvedTargetName, out targetAddress))
+            if (!InitProcessPing(targetNameOrAddress, out string resolvedTargetName, out IPAddress targetAddress))
             {
                 return;
             }
 
-            TraceRouteResult traceRouteResult = new TraceRouteResult(Source, targetAddress, resolvedTargetName);
-
             int currentHop = 1;
             PingOptions pingOptions = new PingOptions(currentHop, DontFragment.IsPresent);
-            PingReply reply = null;
+            PingReply reply;
+            PingReply discoveryReply;
             int timeout = TimeoutSeconds * 1000;
+            Stopwatch timer = new Stopwatch();
 
+            IPAddress hopAddress;
             do
             {
-                TraceRouteReply traceRouteReply = new TraceRouteReply();
-
-                pingOptions.Ttl = traceRouteReply.Hop = currentHop;
-                currentHop++;
-
-                // In the specific case we don't use 'Count' property.
+                // Clear the stored router name for every hop
+                string routerName = null;
+                pingOptions.Ttl = currentHop;
+
+#if !UNIX
+                // Get intermediate hop target. This needs to be done first, so that we can target it properly
+                // and get useful responses.
+                var discoveryAttempts = 0;
+                bool addressIsValid = false;
+                do
+                {
+                    discoveryReply = SendCancellablePing(targetAddress, timeout, buffer, pingOptions);
+                    discoveryAttempts++;
+                    addressIsValid = !(discoveryReply.Address.Equals(IPAddress.Any)
+                        || discoveryReply.Address.Equals(IPAddress.IPv6Any));
+                }
+                while (discoveryAttempts <= 3 && addressIsValid);
+
+                // If we aren't able to get a valid address, just re-target the final destination of the trace.
+                hopAddress = addressIsValid ? discoveryReply.Address : targetAddress;
```

Comment:
Yes, that's what happening here. But if we're getting no valid address from the router at that point (e.g., it gives us a TimedOut rather than TtlExpired) we cannot specifically target the router in question. The TTL value for the current hop point is still used, so we will still target the correct top point, just be unable to target the specific address.

This is critical to allowing traces to work on Unix especially (cannot get TtlExpired response from .net core under Unix, it always returns TimedOut with no valid address in the reply) but can still occur on Windows if the router refuses to respond.

Reply:
We can freely ignore bugs for Buffer and PingOptions.
As for TtlExpired do you discover new bug? Even so, I don’t understand why it should be outside the main loop below.

Reply:
It might be a new one? Not sure. I noticed it after we moved to .NET Core 3, I don't recall if it was different before. 

It's outside the main loop because we don't want to send more pings than necessary. We send one ping to determine address (which is the only useful information we can get out of a TtlExpired response from the current API), and then we have no need to send this "discovery" ping again until we target the next hop. The loop you refer to only executes three times, then returns to the outer loop. 

Or are you saying we should just do the "discovery" ping as part of the inner loop, and then discard the information from the first ping explicitly inside the loop instead of outputting?

Reply:
The more I are thinking about the workaround the more I do not like it.
If we add workarounds here we MUST make the same in all code paths. Otherwise we will confuse users. 
It will complicate our code and output objects a lot and we need to test all the workarounds on all supported platforms.
It's too much effort.
Once Core team fix these bugs we will have to re-write our code and tests. Again it's too much effort.

Reply:
I'm actually not sure this would be considered a bug. I would need to dig pretty deep into the ICMP specifications to determine whether routers are permitted to simply discard buffer information (or other information) when sending a `TtlExpired` response, but intuitively that would make sense, in a way. After all, it's an error state (the packet never reached the intended target).

It's not my favourite solution of all time, and it may not be the optimal one, but it does provide the end user a much better experience than exposing possible bugs from .NET core would.

Reply:
> I would need to dig pretty deep into the ICMP specifications to determine whether routers are permitted to 

Core should cares about this :-)

Reply:
I wish I could say I think they did, but given the amount of weirdness that isn't documented and straight up looks flat out like bugs when working on this, I don't think I can say that I believe they actually do 😓

Reply:
We can honestly tell users that this cmdlet works well on Windows for all features but not on Unix due to bugs in Core.

Reply:
I agree. Even on Windows, though, this particular workaround is necessary for the traceroute's information for each hop point to be accurate.

Reply:
The same :-) If ReplyStatus is broken it is broken always.

Reply:
I think we need to wait Core team conclusion and then make conclusion here.

Reply:
I would like the functionality to be as workable as I can make it. Even if the API is broken, I do not see an advantage in propagating the broken behaviour to users. With the `PingStatus` and `TraceStatus` object types we can mask _some_ of the broken behaviour where we have the capability without extraordinary effort, and document the rest.

When the .NET Core team fix the API we can then simply reduce the overhead on the `PingStatus`/`TraceStatus` objects and have them simply directly reference `PingReply` properties instead.

Reply:
I prefer to get understanding Core team intentions before doing something here.

Reply:
Yeah, I agree that sounds like the best course of action. Thanks!

I'll keep an eye on the coreFX issue, then. 🙂